tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala

```diff
@@ -233,14 +246,63 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
       }
     }
 
-    if (resultType != AnyType) {
-      reportError(s"Result type of exported method def is $resultType, "+
-          "but must be Any")
+    if (classDef.kind == ClassKind.JSClass && name == "constructor") {
+      checkJSClassConstructor(methodDef, classDef)
+    } else {
+      if (resultType != AnyType) {
+        reportError(s"Result type of exported method def is $resultType, "+
+            "but must be Any")
+      }
+
+      val thisType =
+        if (classDef.kind == ClassKind.JSClass) AnyType
+        else ClassType(classDef.name.name)
+
+      val bodyEnv = Env.fromSignature(thisType, params, resultType)
+      typecheckExpect(body, bodyEnv, resultType)
     }
+  }
 
-    val thisType = ClassType(classDef.name.name)
-    val bodyEnv = Env.fromSignature(thisType, params, resultType)
-    typecheckExpect(body, bodyEnv, resultType)
+  def checkJSClassConstructor(methodDef: MethodDef,
+      classDef: LinkedClass): Unit = {
+    val MethodDef(static, _, params, resultType, body) = methodDef
+    implicit val ctx = ErrorContext(methodDef)
+
+    if (resultType != NoType) {
+      reportError(s"Result type of JS class constructor is $resultType, "+
+          "but must be NoType")
+    }
+
+    val bodyStats = body match {
+      case Block(stats) => stats
+      case _            => body :: Nil
+    }
+
+    val (prepStats, superCallAndRest) =
+      bodyStats.span(!_.isInstanceOf[JSSuperConstructorCall])
+
+    val (superCall, restStats) = superCallAndRest match {
+      case (superCall: JSSuperConstructorCall) :: restStats =>
+        (superCall, restStats)
+      case _ =>
+        reportError(
+            "A JS class constructor must contain one super constructor call " +
+            "at the top-level")
+        (JSSuperConstructorCall(Nil)(methodDef.pos), Nil)
+    }
+
+    val initialEnv = Env.fromSignature(NoType, params, NoType,
+        isConstructor = true)
+
+    val preparedEnv = (initialEnv /: prepStats) { (prevEnv, stat) =>
+      typecheckStat(stat, prevEnv)
+    }
+
+    for (arg <- superCall.args)
+      typecheckExprOrSpread(arg, preparedEnv)
```

Comment:
Can io.js parse the file? If yes, let it crash at runtime (there, the flag will work, right?). We say we emit ES6. If the VM does not support ES6, it's not our job.

In other words: I don't agree on the precedent argument, since the culprit here is the VMs lacking support, but I see that testability is an issue...